src/simd/generic.rs

```diff
@@ -1,33 +1,31 @@
+use num_traits::float::FloatCore;
+use num_traits::int::PrimInt;
 use num_traits::AsPrimitive;
 
+use super::config::SIMDInstructionSet;
 use super::task::*;
-use crate::scalar::{ScalarArgMinMax, SCALAR};
+use crate::scalar::{SCALARIgnoreNaN, ScalarArgMinMax, SCALAR};
 
-// TODO: other potential generic SIMDIndexDtype: Copy
-#[allow(clippy::missing_safety_doc)] // TODO: add safety docs?
-pub trait SIMD<
+// ---------------------------------- SIMD operations ----------------------------------
+
+/// Core SIMD operations
+pub trait SIMDOps<ScalarDType, SIMDVecDtype, SIMDMaskDtype, const LANE_SIZE: usize>
+where
     ScalarDType: Copy + PartialOrd + AsPrimitive<usize>,
     SIMDVecDtype: Copy,
     SIMDMaskDtype: Copy,
-    const LANE_SIZE: usize,
->
 {
+    /// Integers > this value **cannot** be accurately represented in SIMDVecDtype
+    const MAX_INDEX: usize;
+    /// Initial index value for the SIMD vector
     const INITIAL_INDEX: SIMDVecDtype;
-    const MAX_INDEX: usize; // Integers > this value **cannot** be accurately represented in SIMDVecDtype
-
-    #[inline(always)]
-    fn _find_largest_lower_multiple_of_lane_size(n: usize) -> usize {
-        n - n % LANE_SIZE
-    }
-
-    // ------------------------------------ SIMD HELPERS --------------------------------------
+    /// Increment value for the SIMD vector
+    const INDEX_INCREMENT: SIMDVecDtype;
 
     unsafe fn _reg_to_arr(reg: SIMDVecDtype) -> [ScalarDType; LANE_SIZE];
 
     unsafe fn _mm_loadu(data: *const ScalarDType) -> SIMDVecDtype;
 
-    unsafe fn _mm_set1(a: usize) -> SIMDVecDtype;
-
     unsafe fn _mm_add(a: SIMDVecDtype, b: SIMDVecDtype) -> SIMDVecDtype;
 
     unsafe fn _mm_cmpgt(a: SIMDVecDtype, b: SIMDVecDtype) -> SIMDMaskDtype;
```

Comment:
Let's add some documentation or at least explain what these functions do.

`cmpgt` = "computer gate?" to the uninitiated.

Reply:
I'll document the various SIMD traits!

---

cmpgt = "compare greater-than" 

For example: https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=cmpgt_ps&techs=SSE_ALL&ig_expand=1139